File: Tarea 1/Sistema de ecuaciones/ecuaciones.pl
use Scalar::Util qw(looks_like_number);
use Data::Types qw(is_int);
use Math::Round qw(nearest);

sub validPosInt{
	my ($num) = @_;
	if( looks_like_number($num) ){
		if( is_int($num) ){
			if($num >= 0){
				return 0;
			}
		}
	}
	print "\tSolo numeros enteros positivos\n";
	return 1;
}

sub llenarSis{
	for($i = 0; $i < $Sfc[0]; $i++){
		for($j = 0; $j < $Sfc[1]; $j++){
			do{
				print "\t[$i][$j] = ";
				$sistem[$i][$j] = <>;
			}
			while( validNum($sistem[$i][$j]) );
		}
	}
	
	return;
}

sub validNum{
	my ($num) = @_;
	if( looks_like_number($num) ){
		return 0;
	}
	print "\tSolo numeros\n";
	return 1;
}

sub printSis{
	for($i = 0; $i < $Sfc[0]; $i++){
		print chr(124)." ";
		for($j = 0; $j < $Sfc[1]; $j++){
			chomp $sistem[$i][$j];
			print nearest(.001, $sistem[$i][$j])."\t";
		}
		print chr(124)."\n";
	}
	print "\n";
	return;
}

sub resolver{
	my $pivant = 1;
	my $pivact;
	my $iter = 0;
	my @trans;
	while($iter < $Sfc[0]){
		$pivact = $sistem[$iter][$iter];
		for($j = 0; $j < $Sfc[1]; $j++){
			$trans[$iter][$j] = $sistem[$iter][$j];
		}
		
		for($i = 0; $i < $Sfc[0]; $i++){
			if($i != $iter){
				$trans[$i][$iter] = 0.0;
			}
			else{
				$trans[$i][$i] = $pivact;
			}
		}
		for($i = 0; $i < $Sfc[0]; $i++){
			if($i != $iter) {
				for($j = 0; $j < $Sfc[1]; $j++){
					if($j != $iter){
						$trans[$i][$j] = (($sistem[$i][$j] * $pivact) - ($sistem[$iter][$j] * $sistem[$i][$iter])) / $pivant;
					}
				}
			}
		}
		for($i = 0; $i < $Sfc[0]; $i++){
			for($j = 0; $j < $Sfc[1]; $j++){
				$sistem[$i][$j] = $trans[$i][$j];
			}
		}
		$pivant = $pivact;
		if($pivant == 0){
			print "\n\tNo existe solucion";
			exit 0;
		}
		$iter++;
	}
	for($i = 0; $i < $Sfc[0]; $i++){
		for($j = 0; $j < $Sfc[1]; $j++){
			$sistem[$i][$j] = $sistem[$i][$j] / $pivact;
		}
	}
	
	return;
}

@Sfc;

print "\nResolver sistema de ecuaciones lineales\n";
$Sfc[0] = -1;
do{{
	if($Sfc[0] > 5){
		print "\tMaximo 5 variables\n";
	}
	elsif($Sfc[0] == 0){
		print "\tSistema sin variables\n";
	}
	print "Cantidad de variables en el sistema (maximo 5): ";
	$Sfc[0] = <>;
}}
while(validPosInt($Sfc[0]) or $Sfc[0] == 0 or $Sfc[0] > 5);

$Sfc[1] = $Sfc[0] + 1;

print "\nLos valores de la ultima columna son las igualdades de cada ecuacion, que es cada fila\n";
print "Cada fila es una ecuacion. Cada columna es un coeficiente de la variable, excepto la ultima, que es la igualdad.\n";
print "Llenar sistema en posicion:\n";
@sistem;
llenarSis();

print "\nSistema original:\n";
printSis();
print "Sistema resuelto:\n";
resolver();
printSis();

File: Tarea 1/Sistema de ecuaciones/ecuaciones.py
import numpy

def validar(num, op):           #validar entrada de números
    if(num == ""):                  #si no se introdujo nada
        print("\tIngresar números")
        return "null"                   #regresar "null"
    copia = num                     #crear copia
    if copia[0] == "-":             #si el primer caracter es un signo de menos
        copia = copia.replace("-","",1) #omitir el signo
    if op == "cel":                 #si se está validando que sea float
        copia = copia.replace(".","",1) #omitir un punto
            
    if copia.isdigit() == 0:       #si no es un número
        if op == "int":             #si se quiere int
            print("Solo números enteros positivos")
        elif op == "cel":           #se se quiere float
            print("Solo números")
            
        return "null"                    #regresar que no es número
    else:                           #si sí es número
        if op == "int":    #si se quiere int
            return int(num)               #regresar string convertido a int
        elif op == "cel":               #si se quiere float
            return float(num)                 #regresar strng convertido a 

S = ["null","null"]         #filas y columnas de la matriz del sistema
print("Resolución de sistemas de ecuaciones lineales")
while S[0] == "null":       #mientras las filas no sean válidas
    print("\nIngresar cantidad de variables del sistema (máximo 5): ", end = "")
    S[0] = input()              #leer filas
    S[0] = validar(S[0],"int")  #validar entrada
    if S[0] == "null":
        continue
    
    if S[0] <= 0:                #si las filas son menores a 1
        print("Sistema sin variables")
        S[0] = "null"
    elif S[0] > 5:
        print("Máximo 5 variables")
        S[0] = "null"
    else:                       #si las filas son mayores o igual a 1
        S[1] = S[0] + 1             #las columnas serán filas+1
    
sistem = [["null" for j in range(S[1])] for i in range(S[0])]   #crear matriz
print("Cada fila es una ecuación. Cada columna es un coeficiente de una variable, excepto la última, que es la igualdad.")
print("Llenar sistema en posición:")
for i in range(S[0]):               #recorrer filas
    for j in range(S[1]):               #recorrer columnas
        sistem[i][j] = "null"           #de momento la celda no es válida
        while sistem[i][j] == "null":   #mientras la celda no sea válida
            print("\t[",i,"][",j,"] = ", sep = "", end = "")
            sistem[i][j] = input()          #leer celda
            sistem[i][j] = validar(sistem[i][j], "cel")#validar celda
print("Sistema original:")
print(numpy.matrix(sistem))     #imprimir matriz
print("\nSistema resuelto:")

pivant = 1      #pivote anterior
pivact = 0      #pivote actual
iterador = 0
trans = [["null" for j in range(S[1])] for i in range(S[0])]    #crear matriz
while iterador < S[0]:      #mientras iterador sea menor al número de filas
    pivact = sistem[iterador][iterador] #pivote actual es el primer elemento del sistema
    for j in range(S[1]):       #recorrer columnas
        trans[iterador][j] = sistem[iterador][j] #copiar fila
    for i in range(S[0]):       #recorrer filas
        if i != iterador:           #si es diferente al iterador
            trans[i][iterador] = 0.0    #hacer 0 a los elementos de la columna por debajo del pivote
        else:                       #si son iguales
            trans[i][i] = pivact        #copiar pivote actual
    for i in range(S[0]):       #recorrer filas
        if i != iterador:           #si es diferente al iterador
            for j in range(S[1]):       #recorrer columnas
                if j != iterador:           #si es diferente al iterador
                    trans[i][j] = ((sistem[i][j] * pivact) - (sistem[iterador][j] * sistem[i][iterador])) / pivant
                                                #definir celda
    for i in range(S[0]):       #recorrer filas
        for j in range(S[1]):       #recorrer columnas
            sistem[i][j] = trans[i][j]  #copiar matriz
    pivant = pivact             #pivote 'anterior' ahora es el pivote 'actual'
    if pivant == 0:             #si el pivote es 0
        print("No existe solución")
        exit()                      #salir
    iterador += 1
for i in range(S[0]):       #recorrer filas
    for j in range(S[1]):       #recorrer columnas
        sistem[i][j] = sistem[i][j] / pivact    #dividir celda entre pivote actual
        if sistem[i][j] == -0:
            sistem[i][j] = 0

print(numpy.matrix(sistem))
